### PR TITLE
The binomial theorem

### PR DESCRIPTION
**Pull request by EgbertRijke**
- Expands the theories of commutative monoids, rings, and commutative rings
- Adds the theories of semirings and commutative semirings
- Proves the binomial theorem for semirings, and concludes it for rings, commutative semirings, commutative rings, and for the natural numbers and the integers.

### CODE REVIEW

File: src/elementary-number-theory/modular-arithmetic-standard-finite-types.lagda.md [hunk 1]
```diff
@@ -690,6 +718,10 @@ is-injective-add-Fin' k x {y} {z} p =
     ( commutative-add-Fin k x y ∙ (p ∙ commutative-add-Fin k z x))
 ```
 
+## More laws
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/modular-arithmetic-standard-finite-types.lagda.md:721]: Is there a purpose for this header?

File: src/group-theory/abelian-groups.lagda.md [hunk 2]
```diff
@@ -47,120 +48,131 @@ is-prop-is-abelian-Group G = is-prop-type-Prop (is-abelian-group-Prop G)
...
+
+  left-inverse-law-add-Ab : (x : type-Ab) → add-Ab (neg-Ab x) x ＝ zero-Ab
+  left-inverse-law-add-Ab = left-inverse-law-mul-Group group-Ab
+
+  right-inverse-law-add-Ab : (x : type-Ab) → add-Ab x (neg-Ab x) ＝ zero-Ab
+  right-inverse-law-add-Ab = right-inverse-law-mul-Group group-Ab
+
+  commutative-add-Ab : (x y : type-Ab) → Id (add-Ab x y) (add-Ab y x)
+  commutative-add-Ab = pr2 A
+
```
Comment (fredrik-bakke) [inline src/group-theory/abelian-groups.lagda.md:122]: `## Properties`

File: src/group-theory/commutative-monoids.lagda.md [hunk 3]
```diff
@@ -46,29 +54,106 @@ module _
...
+    ( mul-Commutative-Monoid (mul-Commutative-Monoid x y) z) ＝
+    ( mul-Commutative-Monoid x (mul-Commutative-Monoid y z))
   associative-mul-Commutative-Monoid =
     associative-mul-Monoid monoid-Commutative-Monoid
 
+  commutative-mul-Commutative-Monoid :
+    (x y : type-Commutative-Monoid) →
+    mul-Commutative-Monoid x y ＝ mul-Commutative-Monoid y x
+  commutative-mul-Commutative-Monoid = pr2 M
+
```
Comment (fredrik-bakke) [inline src/group-theory/commutative-monoids.lagda.md:89]: `## Properties`

File: src/linear-algebra/scalar-multiplication-vectors-on-rings.lagda.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+open import linear-algebra.vectors-on-rings
+
+open import ring-theory.homomorphisms-rings
+open import ring-theory.modules-rings
+open import ring-theory.rings
+```
+
+### Scalar multiplication of vectors on rings
+
+```agda
```
Comment (fredrik-bakke) [inline src/linear-algebra/scalar-multiplication-vectors-on-rings.lagda.md:25]:
```
## Idea
[...]
## Definition
[...]
```

Comment (fredrik-bakke) [reply src/linear-algebra/scalar-multiplication-vectors-on-rings.lagda.md:25]: This is already the title of the file

File: src/linear-algebra/scalar-multiplication-vectors-on-rings.lagda.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    {n : ℕ} (r s : type-Ring R) (v : vec-Ring R n) →
+    scalar-mul-vec-Ring (add-Ring R r s) v ＝
+    add-vec-Ring R (scalar-mul-vec-Ring r v) (scalar-mul-vec-Ring s v)
+  right-distributive-scalar-mul-add-vec-Ring r s empty-vec = refl
+  right-distributive-scalar-mul-add-vec-Ring r s (x ∷ v) =
+    ap-binary _∷_
+      ( right-distributive-mul-add-Ring R r s x)
+      ( right-distributive-scalar-mul-add-vec-Ring r s v)
+```
+
```
Comment (fredrik-bakke) [inline src/linear-algebra/scalar-multiplication-vectors-on-rings.lagda.md:70]: `## Properties`

File: src/linear-algebra/vectors-on-commutative-rings.lagda.md [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+open import group-theory.commutative-monoids
+open import group-theory.monoids
+open import group-theory.semigroups
+
+open import linear-algebra.constant-vectors
+open import linear-algebra.functoriality-vectors
+open import linear-algebra.vectors
+open import linear-algebra.vectors-on-rings
+```
+
```
Comment (fredrik-bakke) [inline src/linear-algebra/vectors-on-commutative-rings.lagda.md:24]: Optional: idea-section

File: src/linear-algebra/vectors-on-commutative-semirings.lagda.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+open import group-theory.commutative-monoids
+open import group-theory.monoids
+open import group-theory.semigroups
+
+open import linear-algebra.constant-vectors
+open import linear-algebra.functoriality-vectors
+open import linear-algebra.vectors
+open import linear-algebra.vectors-on-semirings
+```
+
```
Comment (fredrik-bakke) [inline src/linear-algebra/vectors-on-commutative-semirings.lagda.md:24]: Optional: idea-section

File: src/linear-algebra/vectors-on-rings.lagda.md [hunk 8]
```diff
@@ -150,99 +258,15 @@ module _
...
-  vec-Ring-Group n = vec-Ring-Semigroup R n ,
-    (zero-vec-Ring R , left-unit-law-add-vec-Ring R , right-unit-law-add-vec-Ring R) ,
-    (neg-vec-Ring , left-inverse-law-add-vec-Ring , right-inverse-law-add-vec-Ring)
+  pr1 (vec-Ring-Group n) = vec-Ring-Semigroup R n
+  pr1 (pr1 (pr2 (vec-Ring-Group n))) = zero-vec-Ring R
+  pr1 (pr2 (pr1 (pr2 (vec-Ring-Group n)))) = left-unit-law-add-vec-Ring R
+  pr2 (pr2 (pr1 (pr2 (vec-Ring-Group n)))) = right-unit-law-add-vec-Ring R
+  pr1 (pr2 (pr2 (vec-Ring-Group n))) = neg-vec-Ring
+  pr1 (pr2 (pr2 (pr2 (vec-Ring-Group n)))) = left-inverse-law-add-vec-Ring
+  pr2 (pr2 (pr2 (pr2 (vec-Ring-Group n)))) = right-inverse-law-add-vec-Ring
```
Comment (fredrik-bakke) [inline src/linear-algebra/vectors-on-rings.lagda.md:267]: Would read better as multiple definitions

File: src/linear-algebra/vectors.lagda.md [hunk 9]
```diff
@@ -5,16 +5,33 @@ module linear-algebra.vectors where
...
 open import foundation.universe-levels
+
+open import univalent-combinatorics.involution-standard-finite-types
+open import univalent-combinatorics.standard-finite-types
 ```
+
 ## Idea
 
-An vector of elements of type `A` of length `n` is a list of `n` elements of type `A`.
+There are two equivalent definitions of vectors of length `n`. First, a ** listed vector** of length `n` is a list of `n` elements of type `A`. Secondly, a **functional vector** of length `n` is a map `Fin n → A`. We define both types of vectors and show that they are equivalent.
```
Comment (fredrik-bakke) [inline src/linear-algebra/vectors.lagda.md:30]: `**listed vector**`

File: src/ring-theory/binomial-theorem-rings.lagda.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+open import ring-theory.powers-of-elements-rings
+open import ring-theory.rings
+open import ring-theory.sums-rings
+
+open import univalent-combinatorics.standard-finite-types
+```
+
+## Idea
+
+The binomial theorem in commutative rings asserts that for any two elements `x` and `y` of a commutative ring `R` and any natural number `n`, we have
```
Comment (fredrik-bakke) [inline src/ring-theory/binomial-theorem-rings.lagda.md:32]: rings*

File: src/ring-theory/binomial-theorem-rings.lagda.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    (n : ℕ) (f : functional-vec-Ring R (succ-ℕ n)) (x : type-Ring R) →
+    mul-Ring R (binomial-sum-Ring R n f) x ＝
+    binomial-sum-Ring R n (λ i → mul-Ring R (f i) x)
+  right-distributive-mul-binomial-sum-Ring =
+    right-distributive-mul-binomial-sum-Semiring (semiring-Ring R)
+```
+
+## Theorem
+
+### Binomial theorem for commutative rings
```
Comment (fredrik-bakke) [inline src/ring-theory/binomial-theorem-rings.lagda.md:108]: rings*

File: src/ring-theory/binomial-theorem-semirings.lagda.md [hunk 12]
```diff
@@ -0,0 +1,596 @@
...
+open import ring-theory.powers-of-elements-semirings
+open import ring-theory.semirings
+open import ring-theory.sums-semirings
+
+open import univalent-combinatorics.standard-finite-types
+```
+
+## Idea
+
+The binomial theorem in commutative rings asserts that for any two elements `x` and `y` of a commutative ring `R` and any natural number `n`, we have
```
Comment (fredrik-bakke) [inline src/ring-theory/binomial-theorem-semirings.lagda.md:31]: semirings*

File: src/ring-theory/binomial-theorem-semirings.lagda.md [hunk 13]
```diff
@@ -0,0 +1,596 @@
...
+                      ( power-Semiring R m x)
+                      ( power-Semiring R
+                        ( dist-ℕ (succ-ℕ (succ-ℕ n)) m)
+                        ( y))))
+                ( nat-inr-Fin (succ-ℕ n) i)))))))
+```
+
+## Theorem
+
+### Binomial theorem for commutative rings
```
Comment (fredrik-bakke) [inline src/ring-theory/binomial-theorem-semirings.lagda.md:383]: semirings*

File: src/ring-theory/powers-of-elements-rings.lagda.md [hunk 14]
```diff
@@ -5,16 +5,52 @@ module ring-theory.powers-of-elements-rings where
 
 open import elementary-number-theory.natural-numbers
 
+open import foundation.identity-types
 open import foundation.universe-levels
 
+open import ring-theory.powers-of-elements-semirings
 open import ring-theory.rings
 ```
```
Comment (fredrik-bakke) [inline src/ring-theory/powers-of-elements-rings.lagda.md:14]: Optional: idea-section

File: src/ring-theory/powers-of-elements-semirings.lagda.md [hunk 15]
```diff
@@ -0,0 +1,139 @@
+# Powers of elements in semirings
```
Comment (fredrik-bakke) [inline src/ring-theory/powers-of-elements-semirings.lagda.md:1]: Optional: idea-section

File: src/ring-theory/rings.lagda.md [hunk 16]
```diff
@@ -273,6 +295,25 @@ module _
...
+  semiring-Ring : Semiring l
+  pr1 semiring-Ring = commutative-monoid-Ab (ab-Ring R)
+  pr1 (pr1 (pr2 semiring-Ring)) = has-associative-mul-Ring R
+  pr1 (pr1 (pr2 (pr1 (pr2 semiring-Ring)))) = one-Ring R
+  pr1 (pr2 (pr1 (pr2 (pr1 (pr2 semiring-Ring))))) = left-unit-law-mul-Ring R
+  pr2 (pr2 (pr1 (pr2 (pr1 (pr2 semiring-Ring))))) = right-unit-law-mul-Ring R
+  pr1 (pr2 (pr2 (pr1 (pr2 semiring-Ring)))) = left-distributive-mul-add-Ring R
+  pr2 (pr2 (pr2 (pr1 (pr2 semiring-Ring)))) = right-distributive-mul-add-Ring R
+  pr1 (pr2 (pr2 semiring-Ring)) = left-zero-law-mul-Ring R
+  pr2 (pr2 (pr2 semiring-Ring)) = right-zero-law-mul-Ring R
```
Comment (fredrik-bakke) [inline src/ring-theory/rings.lagda.md:314]: Is it possible to split this one into multiple definitions as well?

File: src/ring-theory/rings.lagda.md [hunk 17]
```diff
@@ -3,6 +3,9 @@
 ```agda
```
Comment (fredrik-bakke) [inline src/ring-theory/rings.lagda.md:3]: This file is missing the `agda` at the start of some of its code blocks, so these parts of the code will not be compiled.

Comment (fredrik-bakke) [reply src/ring-theory/rings.lagda.md:3]: E.g. under "Multiplication in a ring"

File: src/ring-theory/semirings.lagda.md [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+
+  right-unit-law-add-Semiring :
+    (x : type-Semiring R) → Id (add-Semiring R x zero-Semiring) x
+  right-unit-law-add-Semiring =
+    right-unit-law-mul-Commutative-Monoid (commutative-monoid-Semiring R)
+```
+
+### Multiplication in a semiring
+
+```
```
Comment (fredrik-bakke) [inline src/ring-theory/semirings.lagda.md:186]: missing `agda`

File: src/ring-theory/semirings.lagda.md [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+
+  right-unit-law-mul-Semiring :
+    (x : type-Semiring R) → Id (mul-Semiring R x one-Semiring) x
+  right-unit-law-mul-Semiring =
+    right-unit-law-mul-Monoid multiplicative-monoid-Semiring
+```
+
+### Scalar multiplication of semiring elements by natural numbers
+
+```
```
Comment (fredrik-bakke) [inline src/ring-theory/semirings.lagda.md:267]: `agda`

File: src/ring-theory/sums-rings.lagda.md [hunk 20]
```diff
@@ -0,0 +1,163 @@
+# Sums of elements in rings
```
Comment (fredrik-bakke) [inline src/ring-theory/sums-rings.lagda.md:1]: Optional: idea-section

File: src/ring-theory/sums-semirings.lagda.md [hunk 21]
```diff
@@ -0,0 +1,222 @@
+# Sums of elements in semirings
```
Comment (fredrik-bakke) [inline src/ring-theory/sums-semirings.lagda.md:1]: Optional: idea-section

File: src/ring-theory/rings.lagda.md (hunk 17 above)
Comment (fredrik-bakke) [reply src/ring-theory/rings.lagda.md:3]: and Scalar (something something)

**Comment by fredrik-bakke** [comment]
Are you done with this PR?

**Comment by EgbertRijke** [comment]
Yep! You can merge it when you have time

**Pull request merged**